hazelcast-sql/src/test/java/com/hazelcast/sql/impl/expression/ParameterTest.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression;
+
+import com.hazelcast.sql.SqlService;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import com.hazelcast.test.annotation.ParallelJVMTest;
+import com.hazelcast.test.annotation.QuickTest;
+import org.apache.calcite.rel.type.RelDataType;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.sql.impl.type.QueryDataType.BIGINT;
+import static com.hazelcast.sql.impl.type.QueryDataType.INT;
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelJVMTest.class})
+public class ParameterTest extends ExpressionTestBase {
```

Comment:
Similarly to columns and constants - we need exhaustive coverage here. Note that verify doesn't test anything at all here - only one invocation is generated.

1. Tests for every combination of input object type vs expected type: proper exceptions and return values/types should be checked.
1. Test for cases, where type inference is non-trivial. For example, `SELECT ? FROM t`. Currently, it throws an exception, but why? I doubt other databases do the same.

Reply:
This `verify` tests parameter expression in isolation, every other single expression `verify` utilizes parameters during verification. For instance, `PlusTest` verifies: `? + whatever` and `whatever + ?` for every possible combination of types, and for every valid substitution of types it also verifies pretty large number of possible value substitutions.

Regarding `SELECT ? FROM t`, that how Calcite works, if there is no type information to assign a type to a parameter it receives `UNKNOWN` type which is not considered valid by the validator. I think it would be hard to workaround this in that specific standalone parameter case.

Reply:
Can we cast `?` to `ANY/OBJECT` during inference?

Reply:
Yes, we can, but the consequence of that would be `? + ?` reporting `Cannot apply '+' to arguments of type '<OBJECT> + <OBJECT>'` error instead of `Illegal use of dynamic parameter`. Usually, type inference is done in the context of some operator, because only the operator knowns which types it expects, but there is no operator in this case of a standalone parameter. We can assign `OBJECT` type for every parameter of unknown type, but seeing only a parameter we can't tell does it participate as an operand of some operator or not, so we can't skip that `OBJECT` type assignment. Probably there is some way to workaround that, but I haven't found it yet.